core/src/main/java/org/jdbi/v3/core/mapper/reflect/ConstructorMapper.java

```diff
@@ -171,52 +175,76 @@ public T map(ResultSet rs, StatementContext ctx) throws SQLException {
-                            + "Verify that the Java compiler is configured to emit parameter names, "
-                            + "that your result set has the columns expected, or annotate the "
-                            + "parameter names explicitly with @ColumnName",
-                        constructor,
-                        paramName
-                   )));
-
-                final Type type = parameter.getParameterizedType();
-                mappers[i] = ctx.findColumnMapperFor(type)
-                    .map(mapper -> new SingleColumnMapper<>(mapper, columnIndex + 1))
-                    .orElseThrow(() -> new IllegalArgumentException(String.format(
-                        "Could not find column mapper for type '%s' of parameter '%s' for constructor '%s'",
-                        type, paramName, constructor)));
-
-                unmatchedColumns.remove(columnNames.get(columnIndex));
+                final OptionalInt columnIndexOptional = findColumnIndex(paramName, columnNames, columnNameMatchers, () -> debugName(parameter));
+                if (columnIndexOptional.isPresent()) {
+                    final int columnIndex = columnIndexOptional.getAsInt();
+
+                    final Type type = parameter.getParameterizedType();
+                    ColumnMapper<?> mapper = ctx.findColumnMapperFor(type)
+                        .orElseThrow(() -> new IllegalArgumentException(String.format(
+                            "Could not find column mapper for type '%s' of parameter '%s' for constructor '%s'",
+                            type, paramName, constructor)));
+
+                    mappers.add(new SingleColumnMapper<>(mapper, columnIndex + 1));
+
+                    unmatchedColumns.remove(columnNames.get(columnIndex));
+
+                } else {
+                    unmatchedParams.add(paramName);
+                }
+
             } else {
-                String nestedPrefix = prefix + anno.value();
+                final String nestedPrefix = prefix + anno.value();
 
-                mappers[i] = nestedMappers
-                    .computeIfAbsent(parameter, p ->
-                        new ConstructorMapper<>(findConstructorFor(p.getType()), nestedPrefix))
-                    .specialize0(rs, ctx, columnNames, columnNameMatchers, unmatchedColumns);
+                RowMapper<?> mapper = nestedMappers
+                    .computeIfAbsent(parameter, p -> new ConstructorMapper<>(findConstructorFor(p.getType()), nestedPrefix))
+                    .specialize0(rs, ctx, columnNames, columnNameMatchers, unmatchedColumns, true);
+
+                mappers.add(mapper);
+            }
+        }
+
+        if (!unmatchedParams.isEmpty()) {
+
+            // nested objects missing all of their 'normal' params should be set to null
+            if (nested && unmatchedParams.size() == Arrays.stream(parameters)
+                .filter(p -> !p.isAnnotationPresent(Nested.class))
+                .count()) {
```

Comment:
This logic is a bit difficult to reason through, which makes me nervous.

From what I can see, there are three scenarios for nested mappers:

* No constructor parameters matched with any columns - map to null
* All constructor parameters matched to columns - map to object
  * Exception: `@Nested` parameters may map to null and still satisfy this scenario.
* Some but not all parameters matched to columns - throw exception

So here's the algorithm as I see it:
* Introduce a local variable `int matchedParameters = 0;`
* Iterate constructor parameters looking for column matches
  * When a `@Nested` parameter is encountered
    * Clone `columnNames` and pass that to the `specialize0()` call as the `unmatchedColumns` parameter.
    * Upon return from `specialize0()`:
      * Call `unmatchedColumns.retainAll(columnNamesClone)`
      * Increment `matchedParameters` if `columnNamesClone` is not equal to `columnNames`. A simple `size()` lookup is probably sufficient.
  * For non-nested parameters
    * If a column is matched:
      * Increment `matchedParameters`
    * Otherwise
      * Add parameter name to `unmatchedParameters` list.
* If `matchedParameters` is equal to the number of parameters, return the regular mapper.
* If `matchedParameters` is zero, return a null mapper.
* Otherwise, throw an exception

Hmm. I feel like there's a hole in the logic somewhere but I can't put my finger on it.

Reply:
Thanks for the feedback!  The main problem is that, in the @Nested part, if we only increment matchedParameters when ```columnNamesClone.size() != columnNames.size()```, we miss the case where all the nested constructor params were all unmatched, which would mean ```columnNamesClone.size() == columnNames.size()```.  Basically, when I implemented this logic, the ```nestedParametersNotReturned``` and ```nestedPrefixParametersNotReturned``` test cases failed.

What I did to fix it was to always increment matchedParameters counter upon getting the row mapper for the @Nested parameter.  The way I justified it was that, in the recursive call, if the matched parameter count was 0 or all matched, we'd correctly return a row mapper.  If they didn't match correctly, we'd just throw an exception anyways and halt execution.  So, since the recursive call itself is validating the matches, returning from it should always increment the match.  The fact that it returned meant it matched.

Reply:
I'm investigating an alternative approach to this. Will post a link when it's ready to look at.

In short: add a new method `ColumnNameMatcher.columnNameStartsWith(String columnName, String prefix)`, with a default implementation that returns true. The method would be used by reflective mappers, to skip a nested mapping if no columns begin with the prefix in question.

Also: Change `specialize0()` to return an `Optional<RowMapper<T>>` to allow the calling method to decide what to do if no row mapper is available. `specialize()` can throw an exception, whereas `specialize0()` (on nested mappers) can choose to ignore it.

Reply:
For now I think we should exclude ConstructorMapper from this rework. The sanity checking logic for that one is just crazy hard to reason about.